---
title: "tximeta: Working with derived transcriptomes"
author: "Michael Love, Rob Patro"
date: "`r format(Sys.time(), '%m/%d/%Y')`"
output: 
  html_document:
    highlight: tango
---

# This package is in beta 

[See README](https://github.com/mikelove/tximeta/blob/master/README.md)

# Setup

First, to try out `tximeta` you'll need the example data, which is
contained in this GitHub repo. Here we download the whole repo as a
ZIP file.

```{r echo=FALSE}
knitr::opts_chunk$set(cache=TRUE)
```

```{r eval=FALSE}
library(here)
dest <- here("extdata","crohns.zip")
download.file("https://github.com/mikelove/crohns/archive/master.zip", dest, method="wget")
unzip(dest, exdir=here("extdata"))
```

# Trying to import quants against a derived transcriptome

```{r}
library(here)
file <- here("extdata/crohns-master/data/ensembl/SRR1813877/quant.sf.gz")
coldata <- data.frame(sample="1", files=file, names="SRR1813877",
                      stringsAsFactors=FALSE)
se <- tximeta(coldata)
```

# Documenting derived transcriptome for reproducible analysis

Clearly, for an arbitrary transcriptome used to generate the
quantification index, which does not directly match any
transcriptomes from known sources, there is not much that can be
done to populate the metadata during quantification import. However,
we can facilitate the following two cases: 

1) the transcriptome was derived locally from one or more publicly
available sources
2) the transcriptome was derived by another group, and they have
produced and shared a file that links their derived transcriptome
to publicly available sources

`tximeta` offers functionality to assist reproducible analysis in both
of these cases.

In the case of the quantification file above, the transcriptome was
generated locally by downloading the Ensembl cDNA FASTA and GTF for
Homo sapiens, version 90, and subsetting to the transcripts on the
"standard" chromsomes: 1-22, X, Y, MT. The following code reproduces
the production of the transcriptome from publicly available sources:

```{r eval=FALSE}
library(ensembldb)
dbfile <- ensDbFromGtf("Homo_sapiens.GRCh38.90.gtf.gz")
edb <- EnsDb("Homo_sapiens.GRCh38.90.sqlite")
# this is similar to Gencode CHR (except gencode has ~150 additional transcripts in the PAR of Y)
txps <- transcripts(edb, filter=AnnotationFilterList(SeqNameFilter(c(1:22, "X", "Y","MT"))))
library(Biostrings)
# this is missing some of the transcripts in the above
seqs <- readDNAStringSet("Homo_sapiens.GRCh38.cdna.all.fa.gz")
names(seqs) <- sub("(.*?)\\..*","\\1",names(seqs)) # split at '.'
common <- intersect(names(txps), names(seqs)) 
seqs.sub <- seqs[common]
writeXStringSet(seqs.sub, filepath="Homo_sapiens.GRCh38.cdna.std.chroms.fa")
```

To make this quantification reproducible, we make a
`derivedTxome` which records key information about the
source of the transcripts FASTA file, and the location of the
transcripts relative to a genome (GTF), it also records the signature
of the derived transcriptome that was computed by *Salmon* during the
`index` step. 

By default, `derivedTxome` will write out a JSON file
which can be shared with others, linking the signature of the index
with the other metadata, including FASTA and GTF sources. It will
write out to a file with the same name as the `indexDir`, but with
a `.json` extension added. This can be prevented with `write=FALSE`.

```{r}
indexDir <- "Homo_sapiens.GRCh38.cdna.std.chroms_salmon_0.8.2"
fastaFTP <- "ftp://ftp.ensembl.org/pub/release-90/fasta/homo_sapiens/cdna/Homo_sapiens.GRCh38.cdna.all.fa.gz"
gtfFTP <- "ftp://ftp.ensembl.org/pub/release-90/gtf/homo_sapiens/Homo_sapiens.GRCh38.90.gtf.gz"
makeDerivedTxome(indexDir=indexDir,
                 source="Ensembl", organism="Homo sapiens",
                 version="90", genome="GRCh38",
                 fasta=fastaFTP, gtf=gtfFTP)
```

After running `makeDerivedTxome`, the connection between this Salmon
index (and its signature) with the sources is saved for persistent
usage (this uses *BiocFileCache* and the specifics of the cache location
may change as `tximeta` develops).

With use of `tximeta` and a `derivedTxome` -- as with `tximeta` on a
standard, un-filtered transcriptome -- the software figures out if the
remote GTF has been accessed before, and on future calls, it will
simply load the stashed metadata (using *BiocFileCache*).

```{r}
se <- tximeta(coldata)
```

We can see the appropriate metadata is attached (and here, using the
Ensembl-style notation of chromosomes, thanks to *ensembldb*).

```{r}
rowRanges(se)
seqinfo(se)
```

# Clear derivedTxomes (only for demonstration)

```{r}
library(BiocFileCache)
bfc <- BiocFileCache(".")
bfcinfo(bfc)
bfcremove(bfc, bfcquery(bfc, "derivedTxomeDF")$rid)
bfcinfo(bfc)
```

# Loading derivedTxome JSON files

If a collaborator or the Suppmentary Files for a publication 
shares a `derivedTxome` JSON file, we can likewise use `tximeta`
to automatically assemble the relevant metadata. This implies that the
other person has used `tximeta` with the function `makeDerivedTxome`
demonstrated above, pointing to their Salmon index and to the FASTA
and GTF source.

We simply point to the JSON file and use `loadDerivedTxome` and then
the relevant metadata is saved for persistent usage (using
*BiocFileCache*).

```{r}
jsonfile <- "Homo_sapiens.GRCh38.cdna.std.chroms_salmon_0.8.2.json"
loadDerivedTxome(jsonfile)
```

Again, using `tximeta` figures out whether it needs to access the
remote GTF or not, and assembles the appropriate object on the user's
behalf. 

```{r}
se <- tximeta(coldata)
```

# Clear derivedTxomes (only for demonstration)

```{r}
bfc <- BiocFileCache(".")
bfcinfo(bfc)
bfcremove(bfc, bfcquery(bfc, "derivedTxomeDF")$rid)
bfcinfo(bfc)
```

# Session info

```{r}
library(devtools)
session_info()
```
